runtime/lua/vim/shared.lua

```diff
@@ -190,4 +190,98 @@ function vim.pesc(s)
   return s:gsub('[%(%)%.%%%+%-%*%?%[%]%^%$]', '%%%1')
 end
 
+--- Type checking validation function
+---
+--- Examples:
+--- <pre>
+---  validate({ arg={ { 'foo' }, 'table' }})                          --> Nop
+---  validate({ arg={ 1, 'table' } })                                 --> error("'arg' is expected table, got number")
+---  validate({ arg1={ { 'foo' }, 'table' }, arg2={ 1, 'string' } })  --> error("'arg2' is expected string, got number")
+--- </pre>
+---
+---@param ... Table or list of table. That table is "argument_name = { validation_target, type_name (, whether nil is allowed) }".
+--- The following can be used as type_names:
+---     - table or tbl
+---     - string or str
+---     - number or num
+---     - boolean or bool
+---     - function or func
+---     - nil
+---     - thread
+---     - userdata
+function vim.validate(...)
+  local function _type_name(t)
+    if t == 'tbl' or t == 'table' then return 'table'  end
+    if t == 'str' or t == 'string' then return 'string'  end
+    if t == 'num' or t == 'number' then return 'number'  end
+    if t == 'bool' or t == 'boolean' then return 'boolean'  end
+    if t == 'func' or t == 'function' then return 'function' end
+    if t == 'callable' then return 'callable' end
+    if t == 'nil' then return 'nil' end
+    if t == 'thread' or t == 'thread' then return 'thread'  end
+    if t == 'userdata' then return 'userdata' end
+
+    error(string.format("Invalid type name '%s'. See \":help validate\" for more info.", t))
+  end
+  local function _check_type(target, expected_type)
+    if expected_type == 'callable' then
+      return vim.is_callable(target)
+    else
+      return type(target) == expected_type
+    end
+  end
+
+  local args = { ... }
+
+  for _, arg in pairs(args) do
+    for arg_name, v in pairs(arg) do
+      assert(type(arg_name) == 'string',string.format('Expected string, got %s', type(arg_name)))
+      assert(type(v) == 'table', string.format('Expected table, got %s', type(v)))
+
+      local actual_arg_type = type(v[1])
+      local expected_type = _type_name(v[2])
+      local is_optional = v[3] == true
+
+      if is_optional then
+        assert(_check_type(v[1], expected_type) or actual_arg_type == 'nil', string.format("'%s' is expected %s or nil, got %s", arg_name, expected_type, actual_arg_type))
```

Comment:
don't need a separate message for optional types. If they're optional they are simply ignored if missing, otherwise they are type-checked.